### DISCUSSION TITLE
First track name isn't shown in some DAW software

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#191] [Q&A] [answered]
Why isn't the first track name shown in some DAW software?

Currently, I only have access to free software so I'm checking the resulted files with MidiEditor and Cakewalk.
As you can see from the screenshots below, MidiEditor shows the track name correctly, but Cakewalk only shows the second one.
[Screenshot of MidiEditor](https://drive.google.com/file/d/1HOE052BxLcC2zecjxKwHjMSnzyuYT1p4/view?usp=sharing)
[Screenshot of Cakewalk](https://drive.google.com/file/d/13DB_M9ii-5dqogkr1kI9TA9iH2EF7ZsH/view?usp=sharing)

So I tried again with one track, but Cakewalk still can't recognize the name.
You may think it's Cakewalk's bug or something, and you may be right, but I have some other artists' midi files with all the track names.

What's wrong with my code?

PS. I also tried with the time division zero, just in case.
      `new TimedEvent(new SequenceTrackNameEvent("My First Track"), 0)`

PS2.
I added more tracks in the code and all the names were shown except for the first one.
And then I removed the first track, and this time, the NEXT FIRST TRACK's name disappeared.
Again, I removed the track, and the same issue occurred with the next track.
It's always the first track name that is not recognized properly.

```
using Melanchall.DryWetMidi.Common;
using Melanchall.DryWetMidi.Core;
using Melanchall.DryWetMidi.Interaction;
using NoteName = Melanchall.DryWetMidi.MusicTheory.NoteName;



var tempoMap = TempoMap.Create(Tempo.FromBeatsPerMinute(100), new TimeSignature(5, 8));

var firstTrackChunk = new TrackChunk();
var secondTrackChunk = new TrackChunk();

var ch0 = (FourBitNumber)0;
var ch1 = (FourBitNumber)1;



using (var notesManager = firstTrackChunk.ManageNotes())
{
    TimedObjectsCollection<Note> notes = notesManager.Objects;

    var length = LengthConverter.ConvertFrom(
                    new MetricTimeSpan(hours: 0, minutes: 0, seconds: 10),
                    0,
                    tempoMap);

    var note = new Note(NoteName.C, 4, length, 500)
    {
        Channel = ch0,
        Velocity = (SevenBitNumber)95
    };
    notes.Add(note);
}

using (var timedEventManager = firstTrackChunk.ManageTimedEvents())
{
    TimedObjectsCollection<TimedEvent> timedEvents = timedEventManager.Objects;

    timedEvents.Add(new TimedEvent(new SequenceTrackNameEvent("My First Track")));

    var programChangeEvent = new ProgramChangeEvent((SevenBitNumber)3);
    programChangeEvent.Channel = ch0;
    timedEvents.Add(new TimedEvent(programChangeEvent));

    var channelVolume = new ControlChangeEvent((SevenBitNumber)7, (SevenBitNumber)50);
    channelVolume.Channel = ch0;
    timedEvents.Add(new TimedEvent(channelVolume, 0));

    var channelPan = new ControlChangeEvent((SevenBitNumber)10, (SevenBitNumber)0);
    channelPan.Channel = ch0;
    timedEvents.Add(new TimedEvent(channelPan, 0));
}



using (var notesManager = secondTrackChunk.ManageNotes())
{
    TimedObjectsCollection<Note> notes = notesManager.Objects;

    var length = LengthConverter.ConvertFrom(
                    new MetricTimeSpan(hours: 0, minutes: 0, seconds: 10),
                    0,
                    tempoMap);

    var note = new Note(NoteName.C, 4, length, 500)
    {
        Channel = ch1,
        Velocity = (SevenBitNumber)95
    };
    notes.Add(note);
}

using (var timedEventManager = secondTrackChunk.ManageTimedEvents())
{
    TimedObjectsCollection<TimedEvent> timedEvents = timedEventManager.Objects;

    timedEvents.Add(new TimedEvent(new SequenceTrackNameEvent("My Second Track")));

    var programChangeEvent = new ProgramChangeEvent((SevenBitNumber)10);
    programChangeEvent.Channel = ch1;
    timedEvents.Add(new TimedEvent(programChangeEvent));

    var channelVolume = new ControlChangeEvent((SevenBitNumber)7, (SevenBitNumber)127);
    channelVolume.Channel = ch1;
    timedEvents.Add(new TimedEvent(channelVolume, 0));

    var channelPan = new ControlChangeEvent((SevenBitNumber)10, (SevenBitNumber)127);
    channelPan.Channel = ch1;
    timedEvents.Add(new TimedEvent(channelPan, 0));
}



var midiFile = new MidiFile(firstTrackChunk, secondTrackChunk);

midiFile.Write(@"E:\Test\Test.mid", true);
```

**Comment by melanchall** [comment, marked as answer]
> You may think it's Cakewalk's bug

It's definitely its bug.

As for your code – it's OK except it's too complicated and also you don't use created tempo map in the file in fact.

> I have some other artists' midi files with all the track names

So read one of these files with `MidiFile.Read` and compare content of the file with your one.

**Reply by Jenix-Park** [reply]
Nice that there's nothing wrong with the code and the library.
I found the issue while I was writing my real project, so copied part of it, pasted it into a new empty project, and altered them to focus on the issue.
That's why it looks weird and has some mistakes, haha.
Thank you.

**Reply by Jenix-Park** [reply]
Hi again.
I opened the result file in Cakewalk and named the first track as the same "My First Track" as in the code again, and then saved it.
I compared the before and after files, and the result:
[Diff Screenshot](https://drive.google.com/file/d/1oSCtX5xSLNtgbsOl_7T2S4Vgj2N4uGb6/view?usp=sharing)

Don't know what this means, and you may not get it too by looking at the screenshot, so I also attach both files here so that you can see them yourself.
[Before](https://drive.google.com/file/d/1XAjVacgNY-U7vKamr9Hf2Sd7RxDLwO1F/view?usp=sharing)
[After](https://drive.google.com/file/d/1flz6qhOyKLLL4Doe04CaoAwUF6sEC11C/view?usp=sharing)

Forgot to mention this.
The AFTER file is correctly recognized in Cakewalk.

PS. I newly created a MIDI file from Cakewalk with only track names in it.
See [this file](https://drive.google.com/file/d/16vXRpVnkC5JEYSsfi9IyEHlDRYv74rH6/view?usp=sharing) too, if you are interested.

**Reply by Jenix-Park** [reply]
I also created a new file in MidiEditor another free MIDI editor mentioned above, and named two tracks there, saved the file, and then read the file in Cakewalk.
And all the track names are visible.

**Reply by melanchall** [reply]
Thank you very much for the files! I'll check them.

**Reply by melanchall** [reply]
@Jenix-Park Can you please provide me a file created in MidiEditor? I've looked into Before and After files and the file created by DryWetMIDI is absolutely correct, but the file from Cakewalk is strange. I will provide more detailed answer after looking into the file from MidiEditor.

**Reply by Jenix-Park** [reply]
Thank you for taking your valuable time on this, and sorry that I was wrong about the midi file created by MidiEditor.
Perhaps I wasn't careful and selected the file created by Cakewalk instead of MidiEditor's.
I don't have the file because I deleted it right after writing the above comment, so had to create another one again.
But this time, the first track totally disappeared in Cakewalk.
So I added one note event to the first track in MidiEditor and reopened the file in Cakewalk.
And the first track appeared with no track name just like the issue we are talking about.

I don't remember the MIDI files opened with all the track names, but now I guess it was because the first track was empty or something like my MidiEditor test file.
If the file's first track had any MIDI events in it, it would appear with no name, but because it was empty, it was removed completely and all the other track names were shown and it seemed correct.

Anyway, I'm attaching the file by MidiEditor.
It contains two tracks, "My First Track" and "My Second Track", and each of them has one note event only.
[MidiEditor file](https://drive.google.com/file/d/1aMY7cPB41yGld8H1T5CEZlV873-QENt4/view?usp=sharing)

I also think it's Cakewalk's bug, but it's really strange.
Now I'm a Cubase user, but I used to write songs using the software for a long time.
Originally, it has a long history and was paid application.
I know many professionals who used it for their commercial albums.
Maybe a new bug was introduced by the new owner Bandlab Technology recently?
I don't know..

**Reply by Jenix-Park** [reply]
I should have tried to open the files modified/created by Cakewalk in MidiEditor before writing the above comments.

I just tried and noticed unexpected behaviors.
With the "After" link file, three tracks were presented in MidiEditor. This means another track was added just by renaming the first track in Cakewalk.
And when trying to open the new Cakewalk MIDI file with empty two tracks, MidiEditor just crashed.

**Reply by Jenix-Park** [reply]
I finally found the MIDI file again and my guess was correct.
Cakewalk doesn't show the FIRST TRACK WITHOUT ANY NOTE EVENTS even though it contains other timed events.
(It does read and apply all the MIDI data contained in the track such as Time Signature and Tempo Change, by the way.)
In this case, the second track is shown as the first track with the correct track name, and I was deceived by this.

If the first track has one or more note events, then the track is shown without any name.

**Reply by melanchall** [reply]
Well, your guess is correct, yes. I'll just add a little more information about why Cakewalk has such behavior and why it's wrong.

MIDI spec defines that tempo map should be placed in a first track chunk of a MIDI file. Well, let's see into Before and After file.

Before:

![image](https://user-images.githubusercontent.com/26361185/171368329-ace10168-74cb-44f8-b32c-96f771df1437.png)

After:

![image](https://user-images.githubusercontent.com/26361185/171368380-a555497e-3b50-4f39-abeb-57d4a8b0f7f3.png)

As you can see tempo events are stored in first track chunk by both DWM and Cakewalk. But I'm pretty sure that Cekewalk thinks that first track chunk is **for tempo map only**. So it always skip getting the name of the first track chunk thinking there is nothing more than tempo map there.

And that's wrong. MIDI spec doesn't say that first track chunk for tempo map only. Tempo map should be in first chunk, but any other events can be placed there. More than that at nowdays this requirement looks ridiculous. It's not a problem to get a tempo map spread through the file.

As for Cubase – it has correct behavior.

**Reply by Jenix-Park** [reply]
Thanks for this detailed information.
Now I think I get what's going on in Cakewalk.
There is a default non-removable track named "Tempo Track" at the top, and Cakewalk may be taking the information for it from the first track and some problems are happening while doing it as you stated above.

Cakewalk was acquired by different companies until recently, in the meanwhile, it became outdated and finally free software.
So the developers there might have had difficulties maintaining the project properly.
